File: template/ios/ChatView/Core/Models/ChatModels.swift
// MARK: - ChatModels.swift
// Доменные модели чата. Единственная точка истины для всех структур данных.
//
// Добавление нового типа контента:
//   1. Добавить case в MessageContent + вложенный Payload
//   2. Реализовать MessageContentView в MessageContentViews.swift
//   3. Добавить case в MessageContentViewFactory
//   4. Добавить case в MessageSizeCalculator.contentHeight(for:bubbleWidth:)
//
// Reply-архитектура:
//   • ChatMessage.reply (ReplyInfo) — неизменяемый снапшот момента отправки.
//   • ResolvedReply — результат резолвинга в рантайме через messageIndex.
//   • ReplyInfo.staticSnapshot используется только как fallback когда оригинал удалён.

import Foundation

// MARK: - MessageStatus

enum MessageStatus: String {
    case sending   = "sending"
    case sent      = "sent"
    case delivered = "delivered"
    case read      = "read"
}

// MARK: - MessageContent

/// Версионированный sealed-тип контента сообщения.
/// Каждый case самодостаточен и несёт все данные для рендера.
enum MessageContent {
    case text(TextPayload)
    case image(ImagePayload)
    case mixed(TextPayload, ImagePayload)
    case video(VideoPayload)
    case mixedTextVideo(TextPayload, VideoPayload)
    case voice(VoicePayload)
    case poll(PollPayload)
    case file(FilePayload)

    // MARK: Payloads

    struct TextPayload: Equatable {
        let body: String
    }

    struct ImagePayload: Equatable {
        let url:          String
        let width:        CGFloat?
        let height:       CGFloat?
        let thumbnailUrl: String?
    }

    struct VoicePayload: Equatable {
        let url:      String
        let duration: TimeInterval
        let waveform: [CGFloat]
    }

    struct VideoPayload: Equatable {
        let url:          String
        let thumbnailUrl: String?
        let width:        CGFloat?
        let height:       CGFloat?
        let duration:     TimeInterval?
    }

    struct PollOption: Equatable {
        let id:         String
        let text:       String
        let votes:      Int
        let percentage: CGFloat
    }

    struct PollPayload: Equatable {
        let id:                String
        let question:          String
        let options:           [PollOption]
        let totalVotes:        Int
        let selectedOptionIds: [String]
        let isMultipleChoice:  Bool
        let isClosed:          Bool
    }

    struct FilePayload: Equatable {
        let url:      String
        let name:     String
        let size:     Int64
        let mimeType: String?
    }

    // MARK: Convenience accessors

    var text: String? {
        switch self {
        case .text(let p):              return p.body
        case .mixed(let p, _):          return p.body
        case .mixedTextVideo(let p, _): return p.body
        case .image, .video, .voice, .poll, .file: return nil
        }
    }

    var voice: VoicePayload? {
        if case .voice(let p) = self { return p }
        return nil
    }

    var image: ImagePayload? {
        switch self {
        case .image(let p):    return p
        case .mixed(_, let p): return p
        default:               return nil
        }
    }

    var video: VideoPayload? {
        switch self {
        case .video(let p):              return p
        case .mixedTextVideo(_, let p):  return p
        default:                         return nil
        }
    }

    var poll: PollPayload? {
        if case .poll(let p) = self { return p }
        return nil
    }

    var file: FilePayload? {
        if case .file(let p) = self { return p }
        return nil
    }

    var hasText:  Bool { text  != nil }
    var hasImage: Bool { image != nil }
    var hasVideo: Bool { video != nil }
    var hasVoice: Bool { voice != nil }
    var hasPoll:  Bool { poll  != nil }
    var hasFile:  Bool { file  != nil }

    /// true если контент содержит медиа (изображение или видео) — используется для ширины пузыря.
    var hasMedia: Bool { hasImage || hasVideo }
}

extension MessageContent: Equatable {
    static func == (lhs: MessageContent, rhs: MessageContent) -> Bool {
        switch (lhs, rhs) {
        case (.text(let l),           .text(let r)):              return l == r
        case (.image(let l),          .image(let r)):             return l == r
        case (.mixed(let lt, let li), .mixed(let rt, let ri)):    return lt == rt && li == ri
        case (.video(let l),          .video(let r)):             return l == r
        case (.mixedTextVideo(let lt, let lv), .mixedTextVideo(let rt, let rv)):
            return lt == rt && lv == rv
        case (.voice(let l),          .voice(let r)):             return l == r
        case (.poll(let l),           .poll(let r)):              return l == r
        case (.file(let l),           .file(let r)):              return l == r
        default:                                                   return false
        }
    }
}

// MARK: - Reaction

struct Reaction: Equatable {
    let emoji:  String
    let count:  Int
    let isMine: Bool
}

// MARK: - ReplyInfo

struct ReplyInfo: Equatable {
    let replyToId:  String
    let senderName: String?
    let text:       String?
    let hasImage:   Bool
}

// MARK: - ChatMessage

struct ChatMessage: Identifiable, Equatable {
    let id:            String
    let content:       MessageContent
    let timestamp:     Date
    let senderName:    String?
    let isMine:        Bool
    let groupDate:     String
    let status:        MessageStatus
    let reply:         ReplyInfo?
    let forwardedFrom: String?
    let reactions:     [Reaction]
    let isEdited:      Bool
    let actions:       [MessageAction]

    // MARK: Convenience

    var text:      String?                      { content.text }
    var image:     MessageContent.ImagePayload? { content.image }
    var video:     MessageContent.VideoPayload? { content.video }
    var voice:     MessageContent.VoicePayload? { content.voice }
    var hasText:   Bool                         { content.hasText }
    var hasImage:  Bool                         { content.hasImage }
    var hasVideo:  Bool                         { content.hasVideo }
    var hasVoice:  Bool                         { content.hasVoice }
    var replyToId: String?                      { reply?.replyToId }
}

// MARK: - MessageAction

struct MessageAction: Equatable {
    let id:            String
    let title:         String
    let systemImage:   String?
    let isDestructive: Bool
}

// MARK: - MessageActionRepresentable

protocol MessageActionRepresentable {
    var id:            String  { get }
    var title:         String  { get }
    var systemImage:   String? { get }
    var isDestructive: Bool    { get }
}

extension MessageAction:    MessageActionRepresentable {}
extension ContextMenuAction: MessageActionRepresentable {}

// MARK: - MessageSection

struct MessageSection {
    let dateKey:  String          // "yyyy-MM-dd"
    var messages: [ChatMessage]
}

// MARK: - ResolvedReply

enum ResolvedReply {
    case found(ReplyDisplayInfo)
    case deleted
}

// MARK: - ReplyDisplayInfo

struct ReplyDisplayInfo: Equatable {
    let replyToId:  String
    let senderName: String?
    let text:       String?
    let hasImage:   Bool

    init(from message: ChatMessage) {
        replyToId  = message.id
        senderName = message.senderName
        text       = message.text
        hasImage   = message.hasImage
    }

    init(fromSnapshot info: ReplyInfo) {
        replyToId  = info.replyToId
        senderName = info.senderName
        text       = info.text
        hasImage   = info.hasImage
    }
}

File: template/android/app/src/main/java/com/rnapp/rnchatview/ChatModels.kt
package com.rnapp.rnchatview

enum class MessageStatus(val raw: String) {
    SENDING("sending"),
    SENT("sent"),
    DELIVERED("delivered"),
    READ("read");

    companion object {
        fun from(raw: String?): MessageStatus =
            entries.firstOrNull { it.raw == raw } ?: SENT
    }
}

sealed class MessageContent {

    data class TextPayload(val body: String)

    data class ImagePayload(
        val url: String,
        val width: Float? = null,
        val height: Float? = null,
        val thumbnailUrl: String? = null,
    )

    data class VideoPayload(
        val url: String,
        val thumbnailUrl: String? = null,
        val width: Float? = null,
        val height: Float? = null,
        val duration: Double? = null,
    )

    data class PollOption(
        val id: String,
        val text: String,
        val votes: Int = 0,
        val percentage: Float = 0f,
    )

    data class PollPayload(
        val id: String,
        val question: String,
        val options: List<PollOption>,
        val totalVotes: Int = 0,
        val selectedOptionId: String? = null,
        val isClosed: Boolean = false,
    )

    data class VoicePayload(
        val url: String,
        val duration: Double = 0.0,
        val waveform: List<Float> = emptyList(),
    )

    data class FilePayload(
        val url: String,
        val name: String,
        val size: Long = 0,
        val mimeType: String? = null,
    )

    data class Text(val payload: TextPayload) : MessageContent()
    data class Image(val payload: ImagePayload) : MessageContent()
    data class Mixed(val text: TextPayload, val image: ImagePayload) : MessageContent()
    data class Video(val payload: VideoPayload) : MessageContent()
    data class MixedTextVideo(val text: TextPayload, val video: VideoPayload) : MessageContent()
    data class Voice(val payload: VoicePayload) : MessageContent()
    data class Poll(val payload: PollPayload) : MessageContent()
    data class File(val payload: FilePayload) : MessageContent()

    val textBody: String? get() = when (this) {
        is Text -> payload.body
        is Mixed -> text.body
        is MixedTextVideo -> text.body
        is Image, is Video, is Voice, is Poll, is File -> null
    }

    val voicePayload: VoicePayload? get() = when (this) {
        is Voice -> payload
        else -> null
    }

    val imagePayload: ImagePayload? get() = when (this) {
        is Image -> payload
        is Mixed -> image
        else -> null
    }

    val videoPayload: VideoPayload? get() = when (this) {
        is Video -> payload
        is MixedTextVideo -> video
        else -> null
    }

    val pollPayload: PollPayload? get() = when (this) {
        is Poll -> payload
        else -> null
    }

    val filePayload: FilePayload? get() = when (this) {
        is File -> payload
        else -> null
    }

    val hasText: Boolean get() = textBody != null
    val hasImage: Boolean get() = imagePayload != null
    val hasVideo: Boolean get() = videoPayload != null
    val hasVoice: Boolean get() = voicePayload != null
    val hasPoll: Boolean get() = pollPayload != null
    val hasFile: Boolean get() = filePayload != null
    val hasMedia: Boolean get() = hasImage || hasVideo
}

data class Reaction(
    val emoji: String,
    val count: Int,
    val isMine: Boolean = false,
)

data class ReplyInfo(
    val replyToId: String,
    val snapshotSenderName: String? = null,
    val snapshotText: String? = null,
    val snapshotHasImage: Boolean = false,
)

data class ChatMessage(
    val id: String,
    val content: MessageContent,
    val timestamp: Long,
    val senderName: String? = null,
    val isMine: Boolean = false,
    val groupDate: String = "",
    val status: MessageStatus = MessageStatus.SENT,
    val reply: ReplyInfo? = null,
    val forwardedFrom: String? = null,
    val reactions: List<Reaction> = emptyList(),
    val isEdited: Boolean = false,
    val actions: List<MessageAction> = emptyList(),
) {
    val text: String? get() = content.textBody
    val image: MessageContent.ImagePayload? get() = content.imagePayload
    val video: MessageContent.VideoPayload? get() = content.videoPayload
    val voice: MessageContent.VoicePayload? get() = content.voicePayload
    val hasText: Boolean get() = content.hasText
    val hasImage: Boolean get() = content.hasImage
    val hasVideo: Boolean get() = content.hasVideo
    val hasVoice: Boolean get() = content.hasVoice
    val replyToId: String? get() = reply?.replyToId
}

data class MessageAction(
    val id: String,
    val title: String,
    val systemImage: String? = null,
    val isDestructive: Boolean = false,
)

data class MessageSection(
    val dateKey: String,
    val messages: List<ChatMessage>,
)

sealed class ResolvedReply {
    data class Found(val info: ReplyDisplayInfo) : ResolvedReply()
    data class Deleted(val info: ReplyDisplayInfo) : ResolvedReply()
}

data class ReplyDisplayInfo(
    val replyToId: String,
    val senderName: String? = null,
    val text: String? = null,
    val hasImage: Boolean = false,
    val isDeleted: Boolean = false,
) {
    companion object {
        fun fromLive(message: ChatMessage) = ReplyDisplayInfo(
            replyToId = message.id,
            senderName = message.senderName,
            text = message.text,
            hasImage = message.hasImage,
            isDeleted = false,
        )

        fun fromSnapshot(info: ReplyInfo) = ReplyDisplayInfo(
            replyToId = info.replyToId,
            senderName = info.snapshotSenderName,
            text = info.snapshotText,
            hasImage = info.snapshotHasImage,
            isDeleted = true,
        )
    }
}

sealed class ChatInputAction {
    data class Reply(val messageId: String) : ChatInputAction()
    data class Edit(val messageId: String) : ChatInputAction()
    object None : ChatInputAction()

    companion object {
        fun from(type: String?, messageId: String?): ChatInputAction {
            if (messageId.isNullOrBlank()) return None
            return when (type) {
                "reply" -> Reply(messageId)
                "edit" -> Edit(messageId)
                else -> None
            }
        }
    }
}

enum class ChatScrollPosition { TOP, CENTER, BOTTOM }
